flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputOutputFormatVertex.java

```diff
@@ -99,15 +100,16 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
 					((InitializeOnMaster) outputFormat).initializeGlobal(getParallelism());
 				}
 			}
-
+		} catch (Throwable t) {
+			throw ClusterEntryPointExceptionUtils.tryEnrichClusterEntryPointError(t);
```

Comment:
I think catching the exception here and enriching it is too specific. That way we will leave all other places where this can occur out. I believe that a better approach is to do it where most of the code paths originate from. One idea could be the generation of the `JobManagerRunnerImpl` in the `Dispatcher` and its `resultFuture`. This is the point which will be passed by job submissions as well as job recoveries.

Reply:
I moved the OOM handling into `JobManagerRunnerImpl`'s constructor as discussed in one of our recent discussions. But, I want to point out that it's actually out-of-scope for FLINK-18220.